src/contract.rs

```diff
@@ -0,0 +1,412 @@
+            self.total_shares = 0;
+            self.members = HashMap::new();
+        }
+    }
+
+    /// Withdraws the capital of the member
+    /// Requirements:
+    /// * `msg::source()` must be DAO member
+    /// * The member must have sufficient amount of shares
+    /// * The latest proposal the member voted YES must be processed
+    /// Arguments:
+    /// * `amount`: The amount of shares the member would like to withdraw
+    async fn ragequit(&mut self, amount: u128) {
+        if !self.members.contains_key(&msg::source()) {
+            panic!("account is not a DAO member");
+        }
+        let member = self.members.get_mut(&msg::source()).unwrap();
+        if amount > member.shares {
+            panic!("unsufficient shares");
+        }
+        if let Some(proposal_id) = member.highest_index_yes_vote {
+            if let Some(proposal) = self.proposals.get(&proposal_id) {
+                if !proposal.processed {
+                    panic!("cant ragequit until highest index proposal member voted YES on is processed");
+                }
+            }
+        }
+        member.shares = member.shares.saturating_sub(amount);
+        let funds = self.redeemable_funds(amount).await;
+        transfer_tokens(
+            &self.approved_token_program_id,
+            &exec::program_id(),
+            &msg::source(),
+            funds,
+        )
+        .await;
+        self.total_shares = self.total_shares.saturating_sub(amount);
+        msg::reply(
+            DaoEvent::RageQuit {
+                member: msg::source(),
+                amount: funds,
+            },
+            0,
+        )
+        .unwrap();
+    }
+
+    // calculates the funds that the member can redeem based on his shares
+    async fn redeemable_funds(&self, share: u128) -> u128 {
+        let balance = balance(&self.approved_token_program_id, &exec::program_id()).await;
+        (share * balance) / self.total_shares
+    }
+
+    // calculates a share a user can receive for his deposited tokens
+    async fn calculate_share(&self, tokens: u128) -> u128 {
+        let balance = balance(&self.approved_token_program_id, &exec::program_id()).await;
+        if balance == 0 {
+            return tokens;
+        }
+        (self.total_shares * tokens) / balance
+    }
+
+    // checks that account is DAO member
+    fn is_member(&self, account: &ActorId) -> bool {
+        match self.members.get(account) {
+            Some(member) => {
+                if member.shares == 0 {
+                    return false;
+                }
+            }
+            None => {
+                return false;
+            }
+        }
+        true
+    }
+
+    // check that `msg::source()` is either a DAO member or a delegate key
+    fn check_for_membership(&self) {
+        if !self.is_member(&msg::source()) {
+            panic!("account is not a DAO member")
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
```

Comment:
Please no `pub` and `unsafe`
```suggestion
extern "C" fn init() {
```